### PR TITLE
Cache filesystem and git work shared between compose and diff

### PR DESCRIPTION
**Pull request by EricHarvey**
Add composer::index::RepoIndex: a pwd-rooted single-pass directory walk that answers path_exists / list_dirs / topology-dir / inferred- dir queries from a hashmap, replacing the overlapping WalkDirs in composer's make_nodes, discover_functions, is_inferred_dir, and the differ's per-function walk. Prunes well-known noise dirs (.git, node_modules, target, .venv, ...) so the up-front build is proportional to the topology size, not the repo size.

Memoize the inline parent-walk helpers (parent_topology_file, parent_tags_file, find_parent_function_role) in kit via find_parent_file / find_self_or_parent_file: functions sharing a parent infra dir collapse N x 8 path_exists syscalls to 8.

Cache redundant git subprocesses: configurator::Config::new now reuses the cached kit::root() instead of forking its own git rev-parse per topology + per function; composer::version:: current_semver and composer::tag::git_author gain process-wide caches; compiler::FunctionSpec::new defers find_revision until the spec actually references {{version}}.

On a 16-topology / 133-function production topology, tc diff drops from 2.5s wall / 16s sys CPU to ~0.66s wall / 1.5s sys CPU (3.7x wall-clock, ~10x sys), with output byte-identical to baseline.

---

> [!NOTE]
> **Medium Risk**
> Touches core topology discovery and dependency analysis paths; while behavior is intended to be identical (with fallbacks and regression tests), missed indexing/alias edge cases could cause functions or nested topologies to be skipped.
> 
> **Overview**
> Speeds up `tc compose`/`tc diff` by introducing a process-wide, pwd-rooted filesystem index (`composer::index::RepoIndex`) and switching topology/function discovery and assorted existence checks to use index-backed `path_exists`/`list_subdirs`/`descendants_of`, with a `WalkDir` fallback when the target is outside the indexed root.
> 
> Reduces repeated parent-directory scans by adding cached `kit::find_parent_file`/`find_self_or_parent_file` and updating callers (e.g. tags/topology/roles). Cuts redundant git subprocesses by caching `composer::version::current_semver` and `tag::git_author`, using cached `kit::root()` in `Config::new`, and deferring per-function `git log` revision resolution in `compiler::FunctionSpec` unless the spec references `{{version}}` (adds `regex` dependency).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e62280dbf6baca7c62bc20ec0ba0c8bd8e04a942. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lib/compiler/src/spec/function.rs [hunk 1]
```diff
@@ -448,6 +448,19 @@ fn render(s: &str, version: &str) -> String {
...
 
+/// Return true iff `s` references the `{{version}}` template variable.
+/// Lets `load_and_render` skip a per-dir `git log -n 1` fork+exec when
+/// the spec doesn't ask for it. Matches `{{version}}` with any amount
+/// of inner whitespace; false positives are harmless (we just fall
+/// back to the original eager-render behaviour).
+fn references_version(s: &str) -> bool {
+    use regex::Regex;
+    use std::sync::OnceLock;
+    static RE: OnceLock<Regex> = OnceLock::new();
```
Comment (icylisper) [inline lib/compiler/src/spec/function.rs:459]: Don't see the point of a lock here

**Review by icylisper — APPROVED** [review]

File: lib/compiler/src/spec/function.rs (hunk 1 above)
Comment (EricHarvey) [reply lib/compiler/src/spec/function.rs:459]: `OnceLock` here isn't really about locking — it's the standard way to lazily compile a `Regex` once and reuse it across calls. `Regex::new` isn't `const`, so you can't put it in a plain `static`. Without `OnceLock`, we'd recompile the regex on every call to `references_version` (once per function in the topology).

After first init, `get_or_init` is just an atomic load — effectively zero cost. The "lock" only exists to satisfy Rust's `Sync` requirement on statics; it never actually contends in practice.
